corgidrp/l3_to_l4.py

```diff
@@ -1078,7 +1079,317 @@ def subtract_stellar_polarization(input_dataset, system_mueller_matrix_cal, nd_m
+    
+    '''
+    dataset = input_dataset.copy()
+    input_datasets, values = dataset.split_dataset(prihdr_keywords=["PSFREF"])
+    if values != [0,1] and values != [1,0]:
+        raise ValueError("PSFREF keyword must be present in the primary header and be either 0 or 1 for all images")
+    ref_index = values.index(True)
+    mean_ref_dset = combine_subexposures(input_datasets[ref_index], num_frames_per_group=None, collapse="mean", num_frames_scaling=False)
+    # undo any NaN assignments in the image since we FFT below
+    nan_inds = np.where(np.isnan(mean_ref_dset[0].data))
+    if len(nan_inds[0]) > 0:
+        mean_ref_dset[0].data[nan_inds] = np.mean(input_datasets[ref_index].all_data[:,nan_inds[0],nan_inds[1]], axis=0)
+    mean_ref = mean_ref_dset[0].copy()
+    all_data = []
+    all_dq = []
+    all_err = []
+    image_list = []
+    for frame in input_datasets[1-ref_index]:    
+        # compute shift between frame and mean_ref 
+        shift = get_shift_correlation(frame.data, mean_ref.data)
+        # shift mean_ref to be on top of frame data
+        shifted_ref = np.roll(mean_ref.data, (shift[0], shift[1]), axis=(0,1))
+        # rescale wavelengh bands to match
+        ref_col_mean = np.mean(shifted_ref,axis=0)
+        ref_col_mean[ref_col_mean==0] = 1 # prevent div by 0
+        scale = np.mean(frame.data,axis=0)/ref_col_mean
+        shifted_scaled_ref = shifted_ref*scale
+
+        shifted_refdq = np.roll(mean_ref.dq, (shift[0], shift[1]), axis=(0,1))
+        # at this point in the pipeline, the err is mainly shot noise, so multiplying the err is appropriate
+        # shifting may throw off err at the edges of the frame, but those pixels aren't used anyways
+        shifted_scaled_referr = np.roll(mean_ref.err[0], (shift[0], shift[1]), axis=(0,1))*scale
+        # to prevent affecting future steps with error propagated forward from pixels here we don't use: 
+        # crop to the relevant region only
+        x0 = frame.ext_hdr['WV0_X']
+        y0 = frame.ext_hdr['WV0_Y']
+        crop_shift_scale_referr = np.zeros_like(shifted_scaled_referr)
+        crop_shift_scale_referr[y0-35:y0+36, x0-25:x0+26] = shifted_scaled_referr[y0-35:y0+36, x0-25:x0+26]
+        shifted_scaled_referr = crop_shift_scale_referr
+        # subtract the shifted, scaled ref from the frame
+        frame.data -= shifted_scaled_ref
+        # update the dq and err arrays
+        frame.dq = np.bitwise_or.reduce([frame.dq, shifted_refdq], axis=0)
+        frame.add_error_term(shifted_scaled_referr, 'spec ref image err after alignment and matching spec image waveband scale')
+        all_data.append(frame.data)
+        all_dq.append(frame.dq)
+        all_err.append(frame.err)
+        image_list.append(frame)
+
+    out_dataset = data.Dataset(image_list+[mean_ref])
+    history_msg = f'RDI PSF subtraction applied using averaged reference image.  Last Image is shifted, scaled mean ref frame.'
```

Comment:
It would be useful in the history to state with filenames were used to build the reference image.